internal/expr/expr_test.go

```diff
@@ -297,3 +314,144 @@ func FuzzParser(f *testing.F) {
 		parser.Parse(s)
 	})
 }
+
+func (s *ExprSuite) TestValidPrepare(c *C) {
+	testList := []struct {
+		input            string
+		prepareArgs      []any
+		expectedPrepared string
+	}{{
+		"SELECT street FROM t WHERE x = $Address.street",
+		[]any{Address{}},
+		"SELECT street FROM t WHERE x = ?",
+	}, {
+		"SELECT p FROM t WHERE x = $Person.id",
+		[]any{Person{}},
+		"SELECT p FROM t WHERE x = ?",
+	}}
+	for _, test := range testList {
+		parser := expr.NewParser()
+		parsedExpr, _ := parser.Parse(test.input)
+		preparedExpr, err := parsedExpr.Prepare(test.prepareArgs...)
+
+		if err != nil {
+			c.Fatal(err)
+		}
+		c.Assert(preparedExpr.SQL, Equals, test.expectedPrepared)
+	}
+}
+
+func (s *ExprSuite) TestPrepareMismatchedStructName(c *C) {
+	testList := []struct {
+		sql     string
+		structs []any
+	}{{
+		sql:     "SELECT street FROM t WHERE x = $Address.street",
```

Comment:
Hopefully the error message should have enough context for these. I've just split this one into two separate tests so that the test titles give a bit more context too.